File: holons/defect.py
"""Temporal-emergence defect δ^t_τ.

For a fine state β and time τ, the defect compares two coarse descriptions:

- actual:    `coarse(Φ_τ(β))` — coarse-grain the τ-evolved fine state
- predicted: `Φ^+_{τ/period}(coarse(β))` — evolve the initial coarse state
             by Φ^+, which translates each placed object by its
             catalogued velocity scaled to τ

The defect is the cardinality of the symmetric difference of these sets
(the discrete pseudometric of §6, instantiated on the §7 shared host
`O × Z^2`).

Placed objects whose pattern is not in the `velocities` map are treated as
stationary, which is the §7 default for uncatalogued reaction products. To
catalogue a new pattern (a still life, an oscillator, a new spaceship),
add it to the `velocities` map passed in by the caller.
"""
from __future__ import annotations

import numpy as np

from .coarse import CoarseState, coarse_state
from .life import evolve
from .patterns import Pattern, Velocity


def coarse_evolve(
    coarse: CoarseState,
    velocities: dict[Pattern, Velocity],
    n_periods: int,
) -> CoarseState:
    """Apply Φ^+ for `n_periods` coarse ticks to `coarse`.

    Each placed object whose pattern is a key in `velocities` is
    translated by `velocities[pattern] * n_periods`. Placed objects whose
    pattern is not in `velocities` are treated as stationary.
    """
    if n_periods < 0:
        raise ValueError(f"n_periods must be >= 0, got {n_periods}")
    result: set[tuple[Pattern, tuple[int, int]]] = set()
    for pattern, (r, c) in coarse:
        v = velocities.get(pattern, (0, 0))
        result.add(
            (pattern, (r + v[0] * n_periods, c + v[1] * n_periods))
        )
    return frozenset(result)


def defect(
    beta: np.ndarray,
    tau: int,
    velocities: dict[Pattern, Velocity],
    period: int,
) -> int:
    """Compute `δ^t_τ(β)` as the symmetric-difference cardinality.

    `tau` must be a positive multiple of `period` so that every
    catalogued object returns to its phase-0 canonical form at the
    measurement time. This is the §7 simplification; a phase-aware
    extension is straightforward.

    `velocities` maps the phase-0 canonical pattern of each catalogued
    object class to its per-period velocity.
    """
    if tau <= 0:
        raise ValueError(f"tau must be positive, got {tau}")
    if tau % period != 0:
        raise ValueError(
            f"tau must be a multiple of period {period}, got {tau}"
        )
    n_periods = tau // period
    initial_coarse = coarse_state(beta)
    predicted = coarse_evolve(initial_coarse, velocities, n_periods)
    evolved = evolve(beta, tau)
    actual = coarse_state(evolved)
    return len(predicted.symmetric_difference(actual))

File: viz/src/gol/defect.ts
// Temporal-emergence defect for Game of Life (port of holons/defect.py).
// Two coarse descriptions are compared at time tau:
//   actual    = coarseState(evolve(beta, tau))
//   predicted = Phi^+(coarseState(beta))  (catalogued objects translated)
// The symmetric-difference defect counts mismatched (pattern, anchor) keys;
// the matched defect grades displacement below creation/destruction.
import type { LifeField } from "./life";
import { step } from "./life";
import { coarseState, canonicalise } from "./coarse";
import type { PlacedObject } from "./coarse";
import { GLIDER_SE, GLIDER_NW, GLIDER_SE_VELOCITY, GLIDER_NW_VELOCITY } from "./patterns";
import { minCostAssignment } from "./matching";

// Catalogue: canonical pattern key -> per-period velocity.
const VELOCITIES = new Map<string, [number, number]>();
VELOCITIES.set(canonicalise(GLIDER_SE.map(([r, c]) => [r, c])).patternKey, GLIDER_SE_VELOCITY);
VELOCITIES.set(canonicalise(GLIDER_NW.map(([r, c]) => [r, c])).patternKey, GLIDER_NW_VELOCITY);

export interface PredictedObject {
  patternKey: string;
  anchor: [number, number];
}

// Phi^+: translate each catalogued object by velocity * nPeriods; uncatalogued
// objects are treated as stationary.
export function coarseEvolve(coarse: PlacedObject[], nPeriods: number): PredictedObject[] {
  return coarse.map((o) => {
    const v = VELOCITIES.get(o.patternKey) ?? [0, 0];
    return {
      patternKey: o.patternKey,
      anchor: [o.anchor[0] + v[0] * nPeriods, o.anchor[1] + v[1] * nPeriods],
    };
  });
}

function key(o: { patternKey: string; anchor: [number, number] }): string {
  return `${o.patternKey}@${o.anchor[0]},${o.anchor[1]}`;
}

export function evolve(f: LifeField, tau: number): LifeField {
  let r = f;
  for (let t = 0; t < tau; t++) r = step(r);
  return r;
}

export interface DefectResult {
  symdiff: number;
  matched: number;
  predicted: PredictedObject[];
  actual: PlacedObject[];
}

// Both defects at time tau (tau must be a multiple of period for phase alignment).
export function defectAt(beta: LifeField, tau: number, period: number): DefectResult {
  const nPeriods = Math.round(tau / period);
  const initial = coarseState(beta);
  const predicted = coarseEvolve(initial, nPeriods);
  const actual = coarseState(evolve(beta, tau));

  // Symmetric-difference defect over (pattern, anchor) keys.
  const predKeys = new Set(predicted.map(key));
  const actKeys = new Set(actual.map(key));
  let symdiff = 0;
  for (const k of predKeys) if (!actKeys.has(k)) symdiff++;
  for (const k of actKeys) if (!predKeys.has(k)) symdiff++;

  return { symdiff, matched: matchedDistance(predicted, actual), predicted, actual };
}

// Graded distance via min-cost matching (port of matched_distance).
export function matchedDistance(
  predicted: PredictedObject[],
  actual: Array<{ patternKey: string; anchor: [number, number] }>,
  unmatchedCost = 1.0,
  dispRate = 0.25,
  dispCap = 0.5,
): number {
  const m = predicted.length;
  const n = actual.length;
  if (m + n === 0) return 0;
  const big = (m + n) * unmatchedCost + 1.0;
  const size = m + n;
  const cost: number[][] = Array.from({ length: size }, () => new Array<number>(size).fill(big));
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (predicted[i].patternKey === actual[j].patternKey) {
        const d =
          Math.abs(predicted[i].anchor[0] - actual[j].anchor[0]) +
          Math.abs(predicted[i].anchor[1] - actual[j].anchor[1]);
        cost[i][j] = Math.min(dispRate * d, dispCap);
      }
    }
    cost[i][n + i] = unmatchedCost; // predicted i destroyed
  }
  for (let j = 0; j < n; j++) {
    cost[m + j][j] = unmatchedCost; // actual j created
    for (let l = 0; l < m; l++) cost[m + j][n + l] = 0; // ghost filler
  }
  return minCostAssignment(cost);
}
